### DISCUSSION TITLE
PdfReader.outline how to get the page content by title

### DISCUSSION THREAD

**Discussion by Ethan-Chen-plus** [py-pdf/pypdf#2042] [Q&A]
```
from PyPDF2 import PdfReader as pdf_read

#每个书签的索引格式
#{'/Title': '书签名', '/Page': '指向的目标页数', '/Type': '类型'}

directory_str = ''
def bookmark_listhandler(list):
    global directory_str
    for message in list:
        if isinstance(message, dict):
            directory_str += message['/Title'] + '\n'
            # print(message['/Title'])
        else:
            bookmark_listhandler(message)
a = []
with open('10.1007@s10058-019-00220-4.pdf', 'rb') as f:
    pdf = pdf_read(f)
    #检索文档中存在的文本大纲,返回的对象是一个嵌套的列表
    text_outline_list = pdf.outline
    bookmark_listhandler(text_outline_list)
    global a
    a = text_outline_list
    # print(text_outline_list)

with open('context.txt', 'w', encoding='utf-8') as f:
    f.write(directory_str)
directory_str
```
![image](https://github.com/py-pdf/pypdf/assets/84905965/e829a514-e271-43fb-812f-b49a660e1eea)
I can get the directory of the pdf. but how can I get the context of each title?
![image](https://github.com/py-pdf/pypdf/assets/84905965/a6c9c737-96b7-4381-bafd-4ed5e2b8b355)
We can see the title. But I can't see the content of each title.

**Comment by pubpub-zz** [comment]
PyPDF2 is obsolete. please upgrade to pypdf

**Comment by Ethan-Chen-plus** [comment]
@pubpub-zz  OK, thanks.After upgrading, it looks like below:
```python
from pypdf import PdfReader as pdf_read

# 每个书签的索引格式
# {'/Title': '书签名', '/Page': '指向的目标页数', '/Type': '类型'}

directory_str = ""


def bookmark_listhandler(list):
    global directory_str
    for message in list:
        if isinstance(message, dict):
            directory_str += message["/Title"] + "\n"
            # print(message['/Title'])
        else:
            bookmark_listhandler(message)


a = []
with open("./document/Elsver/ikeda2019.pdf", "rb") as f:
    pdf = pdf_read(f)
    # 检索文档中存在的文本大纲,返回的对象是一个嵌套的列表
    text_outline_list = pdf.outline
    bookmark_listhandler(text_outline_list)
    global a
    a = text_outline_list
    # print(text_outline_list)

with open("context.txt", "w", encoding="utf-8") as f:
    f.write(directory_str)
directory_str
``` 
![image](https://github.com/py-pdf/pypdf/assets/84905965/ab6315f8-58b8-4167-95fb-b8868512e8f0)
![image](https://github.com/py-pdf/pypdf/assets/84905965/6cbb6df8-0852-4ac3-9aa9-3d9d9d44959f)
But still I can't get the content of each title.
If I want to get a dict of each title and its content, what should I do?

**Comment by pubpub-zz** [comment]
if you want to access the raw object, the `indirect_reference` is present to access it:
`text_outline_list[0].indirect_reference`

add a call to `.get_object()` to get access to the object.

**Comment by Ethan-Chen-plus** [comment]
@pubpub-zz Thank you.But I mean that I want to get a dict like this:
![image](https://github.com/py-pdf/pypdf/assets/84905965/ad5c731b-8535-411f-ad31-38b23872e7d1)

`text_outline_list[0].indirect_reference.get_object()`can only get this:
![image](https://github.com/py-pdf/pypdf/assets/84905965/13a938ea-779c-4a07-b4fd-bc170b9da224)

**Comment by pubpub-zz** [comment]
you are looking for the text extraction :  
`text_outline_list["/Title"] : text_outline_list["/Page"].extract_text()`
You should put this in a try/except : some outlines may not reference a page

**Reply by Ethan-Chen-plus** [reply]
![image](https://github.com/py-pdf/pypdf/assets/84905965/8cb3985d-d32c-45c3-afcb-ce2872ffc333)
```
from pypdf import PdfReader as pdf_read


directory_str = ''
def bookmark_listhandler(list):
    global directory_str
    for message in list:
        if isinstance(message, dict):
            try:
                print(message['/Title'])
                # print(message['/Page'])
                print(message['/Page'].extract_text())
                print('*'*20)
            except Exception as e:
                print(e)
        else:
            bookmark_listhandler(message)

f = open('10.1007@s10058-019-00220-4.pdf', 'rb')
pdf = pdf_read(f)

text_outline_list = pdf.outline
bookmark_listhandler(text_outline_list)
```
'DictionaryObject' object has no attribute 'extract_text'

**Reply by pubpub-zz** [reply]
for easiness, you will have also to pass the pdf file as parameter:
then try: 
`pdf.pages[pdf.get_page_number(text_outline_list["/Page"])].extract_text()`

**Reply by Ethan-Chen-plus** [reply]
pdf.get_page_number(message['/Page']) can only get the start pages of each title, which also includes other content. I only want to get the things between two titles
this is my test.pdf
![image](https://github.com/py-pdf/pypdf/assets/84905965/803efa24-e89e-490c-9228-fcf29d217e8c)
![image](https://github.com/py-pdf/pypdf/assets/84905965/78168eb7-9312-441c-9feb-2cb04ba58c76)
It only have two pages.
If I use `pdf.pages[pdf.get_page_number(text_outline_list["/Page"])].extract_text()`,each time it will get the whole page of page 1 or page two.However, for example, I just want to get the content which the title include, for "1 Introduction", I only want to get "This is the content of Introduction", not the whole page 1.
![image](https://github.com/py-pdf/pypdf/assets/84905965/040f13d1-eec9-4f30-b838-8b38a0b59298)
![image](https://github.com/py-pdf/pypdf/assets/84905965/a4867a4e-02a5-4169-a1bd-3e68618a1de0)
![image](https://github.com/py-pdf/pypdf/assets/84905965/18d2d6b2-3aac-4ddb-b5fd-a7b0ab1b2bd8)

**Reply by pubpub-zz** [reply]
You are expecting too much from pypdf but also from pdf format: Outlines provide positions. 
So you will have to analyse the full destination to get start position (be carefull about position to the whole page), then you will have to do the same for the next outline then using visitor functions only extracts text within the good range.

All this should be part of your own code
